File: workers/status/src/evidence/service.ts
import { z } from "zod";
import type {
  ProblemDetails,
  TelemetryReference,
} from "@moesegfault/contracts";
import { TelemetryReferenceSchema } from "@moesegfault/contracts";
import {
  QueryTelemetryReferenceRpcRequestSchema,
  TelemetryCapabilitySchema,
  TelemetryQueryAdapterSchema,
  type EvidenceQueryResult,
  type QueryTelemetryReferenceRpcRequest,
  type QueryTelemetryReferenceRpcResult,
} from "../../../../packages/contracts/src/backend-query.js";

import { executeAdapter } from "./adapters.js";
import { parseBackendConfigs, parseTelemetrySecrets } from "./config.js";
import type {
  EvidenceEnvironment,
  ResolvedEvidenceReference,
} from "./types.js";

type Role = "viewer" | "operator" | "admin";

interface EvidenceRow {
  readonly telemetry_reference_id: string;
  readonly kind: string;
  readonly backend_name: string;
  readonly locator_json: string;
  readonly range_start: string | null;
  readonly range_end: string | null;
  readonly service_name: string;
  readonly deployment_id: string;
  readonly correlation_id: string | null;
  readonly trace_id: string | null;
  readonly span_id: string | null;
  readonly expires_at: string | null;
  readonly capabilities_json: string;
  readonly query_adapter: string;
  readonly ui_url_template: string;
  readonly auth_reference: string;
  readonly enabled: number;
  readonly repository_url: string;
  readonly git_commit: string;
}

/** 可注入的时间与 HTTP 依赖，使安全边界可确定测试 / Injectable clock and HTTP dependencies for deterministic boundary tests. */
export interface EvidenceServiceOptions {
  /** 出站 HTTP 实现 / Outbound HTTP implementation. */
  readonly fetch?: typeof fetch;
  /** 当前时间提供器 / Current-time provider. */
  readonly now?: () => Date;
}

/** 创建仅能按已知 ID 查询的 evidence capability / Create an evidence capability that only queries by known reference ID. */
export function createEvidenceService(
  env: EvidenceEnvironment,
  options: EvidenceServiceOptions = {},
): (
  request: QueryTelemetryReferenceRpcRequest,
) => Promise<QueryTelemetryReferenceRpcResult> {
  return (request) => queryTelemetryReference(env, request, options);
}

/**
 * 查询数据库中已存在的 TelemetryReference；永不接受 URL 或临时查询表达式。
 * Query an existing database TelemetryReference; never accepts a URL or ad-hoc expression.
 *
 * @example
 * ```ts
 * const query = createEvidenceService(env);
 * const result = await query({ principal, correlation_id, telemetry_reference_id });
 * ```
 */
export async function queryTelemetryReference(
  env: EvidenceEnvironment,
  raw: unknown,
  options: EvidenceServiceOptions = {},
): Promise<QueryTelemetryReferenceRpcResult> {
  const authorization = authorize(raw);
  if ("problem" in authorization) return authorization;
  const request = authorization.request;
  const resolved = await resolveReference(
    env.DB,
    request.telemetry_reference_id,
    request.correlation_id,
  );
  if (resolved === null)
    return {
      problem: requestProblem(404, "Telemetry reference not found", request),
    };
  if ("problem" in resolved) return { problem: resolved.problem };

  const now = options.now?.() ?? new Date();
  if (
    resolved.reference.expires_at !== undefined &&
    Date.parse(resolved.reference.expires_at) <= now.getTime()
  ) {
    return {
      data: terminal(
        resolved,
        "expired",
        now,
        "Telemetry retention has expired",
      ),
    };
  }
  if (!resolved.enabled)
    return {
      data: terminal(
        resolved,
        "unavailable",
        now,
        "Telemetry backend is disabled",
      ),
    };
  if (!resolved.capabilities.includes(resolved.reference.kind))
    return {
      data: terminal(
        resolved,
        "unsupported",
        now,
        "Backend does not declare this capability",
      ),
    };
  const adapter = TelemetryQueryAdapterSchema.safeParse(resolved.adapterText);
  if (!adapter.success)
    return {
      data: terminal(
        resolved,
        "unsupported",
        now,
        "Backend query adapter is not supported",
      ),
    };
  if (!adapterMatches(adapter.data, resolved.reference.kind))
    return {
      data: terminal(
        resolved,
        "unsupported",
        now,
        "Adapter cannot query this evidence kind",
      ),
    };

  if (
    resolved.reference.kind === "artifact" &&
    !(await artifactExists(env.DB, resolved.reference))
  ) {
    return {
      data: terminal(
        resolved,
        "unavailable",
        now,
        "Artifact is not registered for this deployment",
      ),
    };
  }
  const configs = parseBackendConfigs(env.TELEMETRY_BACKEND_CONFIG_JSON);
  const config = configs[resolved.reference.backend];
  if (config === undefined)
    return {
      data: terminal(
        resolved,
        "unavailable",
        now,
        "Backend runtime configuration is unavailable",
      ),
    };
  const secrets = parseTelemetrySecrets(env.TELEMETRY_AUTH_JSON);
  let result;
  try {
    result = await executeAdapter({
      resolved,
      adapter: adapter.data,
      config,
      credential: secrets[resolved.authReference],
      fetch: options.fetch ?? fetch,
      now,
    });
  } catch {
    return {
      data: terminal(
        resolved,
        "unavailable",
        now,
        "Evidence adapter rejected unsafe or invalid backend data",
      ),
    };
  }
  return {
    data: {
      telemetry_reference: resolved.reference,
      ...result,
      queried_at: now.toISOString(),
    },
  };
}

/** 向后便捷别名 / Convenience alias. */
export const queryEvidence = queryTelemetryReference;

async function resolveReference(
  database: D1Database,
  id: string,
  correlationId: string,
): Promise<
  ResolvedEvidenceReference | { readonly problem: ProblemDetails } | null
> {
  const row = await database
    .prepare(
      `SELECT t.*, b.capabilities_json, b.query_adapter, b.ui_url_template,
            b.auth_reference, b.enabled, d.repository_url, d.git_commit
       FROM telemetry_references AS t
       JOIN telemetry_backends AS b ON b.backend_name=t.backend_name
       JOIN deployments AS d ON d.deployment_id=t.deployment_id
      WHERE t.telemetry_reference_id=?`,
    )
    .bind(id)
    .first<EvidenceRow>();
  if (row === null) return null;
  try {
    const locator = JSON.parse(row.locator_json) as unknown;
    const common = {
      id: row.telemetry_reference_id,
      backend: row.backend_name,
      locator,
      service_name: row.service_name,
      deployment_id: row.deployment_id,
      ...(row.correlation_id === null
        ? {}
        : { correlation_id: row.correlation_id }),
      ...(row.trace_id === null ? {} : { trace_id: row.trace_id }),
      ...(row.span_id === null ? {} : { span_id: row.span_id }),
      ...(row.expires_at === null ? {} : { expires_at: row.expires_at }),
    };
    const candidate =
      row.kind === "log_query" ||
      row.kind === "metric_query" ||
      row.kind === "profile"
        ? {
            ...common,
            kind: row.kind,
            time_range: { start: row.range_start, end: row.range_end },
          }
        : { ...common, kind: row.kind };
    const reference = TelemetryReferenceSchema.parse(candidate);
    const capabilities = z
      .array(TelemetryCapabilitySchema)
      .max(6)
      .parse(JSON.parse(row.capabilities_json));
    return {
      reference,
      adapterText: row.query_adapter,
      capabilities,
      uiUrlTemplate: row.ui_url_template,
      authReference: row.auth_reference,
      enabled: row.enabled === 1,
      repositoryUrl: row.repository_url,
      gitCommit: row.git_commit,
    };
  } catch {
    return {
      problem: {
        type: "https://status.moesegfault.dev/problems/dependency-unavailable",
        title: "Stored telemetry reference is invalid",
        status: 503,
        correlation_id: correlationId,
      },
    };
  }
}

async function artifactExists(
  database: D1Database,
  reference: TelemetryReference,
): Promise<boolean> {
  if (reference.kind !== "artifact") return false;
  const locator = reference.locator;
  const found = await database
    .prepare(
      `SELECT artifact_id FROM deployment_artifacts
      WHERE deployment_id=? AND artifact_digest=? AND kind=?
        AND (? IS NULL OR build_id=?) LIMIT 1`,
    )
    .bind(
      reference.deployment_id,
      locator.artifact_digest,
      locator.artifact_kind,
      locator.build_id ?? null,
      locator.build_id ?? null,
    )
    .first<string>("artifact_id");
  return found !== null;
}

function authorize(
  raw: unknown,
):
  | { readonly request: QueryTelemetryReferenceRpcRequest }
  | { readonly problem: ProblemDetails } {
  const parsed = QueryTelemetryReferenceRpcRequestSchema.safeParse(raw);
  const candidate = raw as { correlation_id?: string } | null;
  const correlation =
    typeof candidate?.correlation_id === "string" &&
    /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/u.test(
      candidate.correlation_id,
    )
      ? candidate.correlation_id
      : "018f0000-0000-7000-8000-000000000000";
  if (!parsed.success)
    return {
      problem: {
        type: "https://status.moesegfault.dev/problems/invalid-request",
        title: "Invalid evidence query request",
        status: 400,
        detail: parsed.error.issues
          .map((issue) => issue.message)
          .join("; ")
          .slice(0, 4096),
        correlation_id: correlation,
      },
    };
  if (!hasRole(parsed.data.principal.roles, "viewer"))
    return {
      problem: {
        type: "https://status.moesegfault.dev/problems/forbidden",
        title: "Forbidden",
        status: 403,
        detail: "Role viewer, operator, or admin is required.",
        correlation_id: parsed.data.correlation_id,
      },
    };
  return { request: parsed.data };
}

function hasRole(roles: readonly Role[], required: Role): boolean {
  const rank: Record<Role, number> = { viewer: 1, operator: 2, admin: 3 };
  return roles.some((role) => rank[role] >= rank[required]);
}

function adapterMatches(
  adapter: z.infer<typeof TelemetryQueryAdapterSchema>,
  kind: TelemetryReference["kind"],
): boolean {
  return (
    (
      {
        tempo: "trace",
        loki: "log_query",
        pyroscope: "profile",
        prometheus: "metric_query",
        "source-commit": "source",
        "artifact-registry": "artifact",
      } as const
    )[adapter] === kind
  );
}

function terminal(
  resolved: ResolvedEvidenceReference,
  status: EvidenceQueryResult["status"],
  now: Date,
  detail: string,
): EvidenceQueryResult {
  return {
    telemetry_reference: resolved.reference,
    status,
    ui_url: null,
    records: [],
    truncated: false,
    queried_at: now.toISOString(),
    detail,
  };
}

function requestProblem(
  status: 404,
  title: string,
  request: QueryTelemetryReferenceRpcRequest,
): ProblemDetails {
  return {
    type: "https://status.moesegfault.dev/problems/not-found",
    title,
    status,
    instance: "/rpc/queryTelemetryReference",
    correlation_id: request.correlation_id,
  };
}

File: crates/status-backend/src/evidence/service.rs
//! D1 注册项驱动真实 Workers Fetch，凭据不进入 RPC。 / D1 registrations drive real Workers Fetch; credentials never enter RPC.
use super::{
    config::{self, Config},
    map_reference, normalize, query,
};
use crate::{
    access::AdminRole,
    admin::{RpcContext, RpcProblem},
    database::{Database, Query, SqlValue},
    wire::Id,
};
use futures_util::{
    future::{select, Either},
    StreamExt,
};
use serde_json::{json, Value};
use std::time::Duration;
use url::Url;
use worker::{AbortController, Delay, Env, Fetch, Request, RequestInit, RequestRedirect};
/// 查询已注册 ID 的遥测；调用方必须是可信私有 RPC。 / Query telemetry by registered ID; callers must be trusted private RPC.
///
/// ```ignore
/// let result = query_telemetry_reference(&env, request).await?;
/// ```
pub async fn query_telemetry_reference(env: &Env, raw: Value) -> worker::Result<Value> {
    let operation = "queryTelemetryReference";
    let context = match RpcContext::parse(&raw, &["telemetry_reference_id"]) {
        Ok(c) => c,
        Err(p) => return Ok(json!({"problem":p})),
    };
    if !context.require(AdminRole::Viewer) {
        return Ok(json!({"problem":RpcProblem::new(403,"Forbidden",&raw,operation)}));
    }
    let Some(id) = raw["telemetry_reference_id"]
        .as_str()
        .filter(|s| Id::new((*s).into()).is_ok())
    else {
        return Ok(
            json!({"problem":RpcProblem::new(400,"Invalid evidence query request",&raw,operation)}),
        );
    };
    let db = Database::new(env.d1("DB")?);
    let row=db.first::<Value>(&Query::new("SELECT t.*,b.capabilities_json,b.query_adapter,b.ui_url_template,b.auth_reference,b.enabled,d.repository_url,d.git_commit FROM telemetry_references t JOIN telemetry_backends b ON b.backend_name=t.backend_name JOIN deployments d ON d.deployment_id=t.deployment_id WHERE t.telemetry_reference_id=?",vec![id.into()])).await.map_err(|_|worker::Error::RustError("Evidence database unavailable".into()))?;
    let Some(row) = row else {
        return Ok(
            json!({"problem":RpcProblem::new(404,"Telemetry reference not found",&raw,operation)}),
        );
    };
    let reference = match map_reference(&row) {
        Ok(r) => r,
        Err(_) => {
            return Ok(
                json!({"problem":RpcProblem::new(503,"Stored telemetry reference is invalid",&raw,operation)}),
            )
        }
    };
    let now = js_sys::Date::now() as i64;
    let mut result = execute(env, &db, &row, &reference, now).await;
    result["telemetry_reference"] = reference;
    result["queried_at"] = json!(normalize::timestamp(now));
    Ok(json!({"data":result}))
}
/// 返回终态，不泄漏供应商响应。 / Return a terminal state without leaking vendor responses.
fn terminal(status: &str, detail: &str) -> Value {
    json!({"status":status,"detail":detail,"ui_url":null,"records":[],"truncated":false})
}
/// 依次检查保留期、能力和来源。 / Check retention, capability and provenance in order.
async fn execute(env: &Env, db: &Database, row: &Value, r: &Value, now: i64) -> Value {
    if r["expires_at"]
        .as_str()
        .and_then(epoch)
        .is_some_and(|t| t <= now)
    {
        return terminal("expired", "Telemetry retention has expired");
    }
    if row["enabled"] != 1 {
        return terminal("unavailable", "Telemetry backend is disabled");
    }
    let capabilities = row["capabilities_json"]
        .as_str()
        .and_then(|s| serde_json::from_str::<Vec<String>>(s).ok());
    if capabilities.as_ref().is_none_or(|v| {
        v.len() > 6
            || v.iter().any(|k| {
                ![
                    "trace",
                    "log_query",
                    "metric_query",
                    "profile",
                    "source",
                    "artifact",
                ]
                .contains(&k.as_str())
            })
    }) {
        return terminal("unavailable", "Stored backend capabilities are invalid");
    }
    if !capabilities
        .unwrap_or_default()
        .iter()
        .any(|k| Some(k.as_str()) == r["kind"].as_str())
    {
        return terminal("unsupported", "Backend does not declare this capability");
    }
    let adapter = row["query_adapter"].as_str().unwrap_or("");
    let expected = match adapter {
        "tempo" => "trace",
        "loki" => "log_query",
        "prometheus" => "metric_query",
        "pyroscope" => "profile",
        "source-commit" => "source",
        "artifact-registry" => "artifact",
        _ => return terminal("unsupported", "Backend query adapter is not supported"),
    };
    if r["kind"] != expected {
        return terminal("unsupported", "Adapter cannot query this evidence kind");
    }
    if adapter == "artifact-registry" {
        let l = &r["locator"];
        let build = l["build_id"]
            .as_str()
            .map(SqlValue::from)
            .unwrap_or(SqlValue::Null);
        let found=db.first::<Value>(&Query::new("SELECT artifact_id FROM deployment_artifacts WHERE deployment_id=? AND artifact_digest=? AND kind=? AND (? IS NULL OR build_id=?) LIMIT 1",vec![r["deployment_id"].as_str().unwrap_or("").into(),l["artifact_digest"].as_str().unwrap_or("").into(),l["artifact_kind"].as_str().unwrap_or("").into(),build.clone(),build])).await;
        if !matches!(found, Ok(Some(_))) {
            return terminal(
                "unavailable",
                "Artifact is not registered for this deployment",
            );
        }
    }
    let raw_config = env
        .var("TELEMETRY_BACKEND_CONFIG_JSON")
        .ok()
        .map(|v| v.to_string());
    let configs = config::configs(raw_config.as_deref());
    let Some(c) = configs.get(r["backend"].as_str().unwrap_or("")) else {
        return terminal(
            "unavailable",
            "Backend runtime configuration is unavailable",
        );
    };
    if adapter == "source-commit" {
        return source(row, r, c);
    }
    if adapter == "artifact-registry" {
        let l = &r["locator"];
        let mut attrs = json!({"artifact_digest":l["artifact_digest"]});
        if let Some(build) = l.get("build_id") {
            attrs["build_id"] = build.clone();
        }
        return json!({"status":"ok","ui_url":ui(row,c,l),"records":[normalize::record(l["artifact_kind"].as_str().unwrap_or("artifact"),None,&attrs)],"truncated":false});
    }
    let Some((url, values)) = request_url(c, r, adapter) else {
        return terminal(
            "unsupported",
            "Locator requires finite structured keys with matching service and deployment identity",
        );
    };
    let secrets_raw = env
        .secret("TELEMETRY_AUTH_JSON")
        .ok()
        .map(|v| v.to_string());
    let secrets = config::secrets(secrets_raw.as_deref());
    let credential = secrets.get(row["auth_reference"].as_str().unwrap_or(""));
    match fetch_json(url, c, credential).await {
        Ok(data) => {
            let (records, truncated) = normalize::normalize(adapter, r, &data);
            json!({"status":"ok","ui_url":ui(row,c,&values),"records":records,"truncated":truncated})
        }
        Err((status, detail)) => {
            let mut result = terminal(status, detail);
            if status == "not_found" {
                result["ui_url"] = json!(ui(row, c, &json!({})));
                result.as_object_mut().map(|m| m.remove("detail"));
            }
            result
        }
    }
}
/// 模板插值仅编码值，主机仍精确许可。 / Template substitutions encode values and still enforce exact hosts.
fn ui(row: &Value, c: &Config, values: &Value) -> Option<String> {
    let mut rendered = row["ui_url_template"].as_str()?.to_string();
    let mut unused = Vec::new();
    for (k, v) in values.as_object()? {
        let Some(v) = v.as_str() else { continue };
        let placeholder = format!("{{{k}}}");
        if rendered.contains(&placeholder) {
            rendered = rendered.replace(&placeholder, &encode(v));
        } else {
            unused.push((k, v));
        }
    }
    if rendered.contains(['{', '}']) {
        return None;
    }
    let mut u = config::allowed(&rendered, c)?;
    for (k, v) in unused {
        u.query_pairs_mut().append_pair(k, v);
    }
    let s = u.to_string();
    (s.len() <= 2048).then_some(s)
}
/// 来源必须匹配不可变 deployment。 / Source must match the immutable deployment.
fn source(row: &Value, r: &Value, c: &Config) -> Value {
    let l = &r["locator"];
    let repository = l["repository_url"].as_str().unwrap_or("");
    if canonical(repository) != canonical(row["repository_url"].as_str().unwrap_or(""))
        || l["git_commit"] != row["git_commit"]
    {
        return terminal(
            "unavailable",
            "Source locator does not match immutable deployment provenance",
        );
    }
    let Some(mut u) = config::allowed(repository, c) else {
        return terminal(
            "unavailable",
            "Source URL is outside the configured HTTPS host allowlist",
        );
    };
    let prefix = u
        .path()
        .trim_end_matches('/')
        .trim_end_matches(".git")
        .to_string();
    let blob = if u.host_str() == Some("gitlab.com") {
        "/-/blob/"
    } else {
        "/blob/"
    };
    let path = l["path"]
        .as_str()
        .unwrap_or("")
        .split('/')
        .map(encode)
        .collect::<Vec<_>>()
        .join("/");
    u.set_path(&format!(
        "{prefix}{blob}{}/{path}",
        l["git_commit"].as_str().unwrap_or("")
    ));
    u.set_query(None);
    u.set_fragment(l["line"].as_u64().map(|n| format!("L{n}")).as_deref());
    if u.as_str().len() > 2048 {
        return terminal("unavailable", "Source URL exceeds the safe length limit");
    }
    let mut attrs = json!({"git_commit":l["git_commit"]});
    for key in ["line", "column"] {
        if let Some(v) = l.get(key) {
            attrs[key] = v.clone();
        }
    }
    json!({"status":"ok","ui_url":u.as_str(),"records":[normalize::record(l["path"].as_str().unwrap_or("source"),None,&attrs)],"truncated":false})
}
fn canonical(raw: &str) -> Option<String> {
    let u = Url::parse(raw).ok()?;
    Some(format!(
        "{}{}",
        u.origin().ascii_serialization(),
        u.path().trim_end_matches('/').trim_end_matches(".git")
    ))
}
fn encode(s: &str) -> String {
    const SET: &percent_encoding::AsciiSet = &percent_encoding::NON_ALPHANUMERIC
        .remove(b'-')
        .remove(b'_')
        .remove(b'.')
        .remove(b'!')
        .remove(b'~')
        .remove(b'*')
        .remove(b'\'')
        .remove(b'(')
        .remove(b')');
    percent_encoding::utf8_percent_encode(s, SET).to_string()
}
fn epoch(s: &str) -> Option<i64> {
    chrono::DateTime::parse_from_rfc3339(s)
        .ok()
        .map(|d| d.timestamp_millis())
}
/// URL 根只来自部署配置。 / URL roots come exclusively from deployment config.
fn request_url(c: &Config, r: &Value, adapter: &str) -> Option<(Url, Value)> {
    let mut u = config::allowed(&c.endpoint, c)?;
    let start = r["time_range"]["start"].as_str().unwrap_or("");
    let end = r["time_range"]["end"].as_str().unwrap_or("");
    let path = match adapter {
        "tempo" => format!("api/v2/traces/{}", r["locator"]["trace_id"].as_str()?),
        "loki" => "loki/api/v1/query_range".into(),
        "prometheus" => "api/v1/query_range".into(),
        "pyroscope" => "pyroscope/render".into(),
        _ => return None,
    };
    u.set_path(&format!("{}/{path}", u.path().trim_end_matches('/')));
    u.set_query(None);
    u.set_fragment(None);
    if adapter == "tempo" {
        return Some((u, json!({"trace_id":r["locator"]["trace_id"]})));
    }
    let expression = query::compile(r)?;
    let start_ms = epoch(start)?;
    let end_ms = epoch(end)?;
    {
        let mut q = u.query_pairs_mut();
        q.append_pair("query", &expression);
        match adapter {
            "loki" => {
                q.append_pair("start", &format!("{start_ms}000000"))
                    .append_pair("end", &format!("{end_ms}000000"))
                    .append_pair("limit", "100")
                    .append_pair("direction", "backward");
            }
            "prometheus" => {
                q.append_pair("start", start)
                    .append_pair("end", end)
                    .append_pair(
                        "step",
                        &(((end_ms - start_ms).max(1000) + 99999) / 100000)
                            .max(1)
                            .to_string(),
                    );
            }
            "pyroscope" => {
                q.append_pair("from", &(start_ms.div_euclid(1000)).to_string())
                    .append_pair("until", &(end_ms.div_euclid(1000)).to_string())
                    .append_pair("format", "json")
                    .append_pair("maxNodes", "100");
            }
            _ => return None,
        }
    }
    let values = if adapter == "pyroscope" {
        json!({"query":expression,"from":start,"until":end})
    } else {
        json!({"query":expression,"start":start,"end":end})
    };
    Some((u, values))
}
/// 截止时间覆盖流式正文；失败总是取消网络。 / Deadline covers streaming body; failure always cancels the network.
async fn fetch_json(
    url: Url,
    c: &Config,
    credential: Option<&String>,
) -> Result<Value, (&'static str, &'static str)> {
    let unavailable = || {
        (
            "unavailable",
            "Telemetry backend request failed or timed out",
        )
    };
    let mut init = RequestInit::new();
    // Workers只支持manual/follow；手动拒绝重定向避免凭据跨主机泄漏。
    // Workers supports manual/follow; rejecting redirects prevents cross-host credential leakage.
    init.with_redirect(RequestRedirect::Manual);
    init.headers
        .set("accept", "application/json")
        .map_err(|_| unavailable())?;
    if c.auth_scheme != "none" {
        let secret = credential.ok_or((
            "unavailable",
            "Configured auth_reference has no secret value",
        ))?;
        init.headers
            .set(
                "authorization",
                &format!(
                    "{} {secret}",
                    if c.auth_scheme == "basic" {
                        "Basic"
                    } else {
                        "Bearer"
                    }
                ),
            )
            .map_err(|_| unavailable())?;
    }
    if let Some(tenant) = &c.tenant_id {
        init.headers
            .set("x-scope-orgid", tenant)
            .map_err(|_| unavailable())?;
    }
    let req = Request::new_with_init(url.as_str(), &init).map_err(|_| unavailable())?;
    let controller = AbortController::default();
    let signal = controller.signal();
    let fetch = async {
        let mut response = Fetch::Request(req)
            .send_with_signal(&signal)
            .await
            .map_err(|_| unavailable())?;
        if (300..400).contains(&response.status_code()) {
            return Err((
                "unavailable",
                "Telemetry backend redirects are not permitted",
            ));
        }
        if response.status_code() == 404 {
            return Err(("not_found", ""));
        }
        if !(200..300).contains(&response.status_code()) {
            return Err((
                "unavailable",
                "Telemetry backend returned an unsuccessful HTTP status",
            ));
        }
        if response
            .headers()
            .get("content-length")
            .ok()
            .flatten()
            .and_then(|s| s.parse::<usize>().ok())
            .is_some_and(|n| n > c.max_response_bytes)
        {
            return Err((
                "unavailable",
                "Telemetry response exceeds the configured byte limit",
            ));
        }
        let mut stream = response.stream().map_err(|_| unavailable())?;
        let mut bytes = Vec::new();
        while let Some(chunk) = stream.next().await {
            let chunk = chunk.map_err(|_| unavailable())?;
            if chunk.len() > c.max_response_bytes.saturating_sub(bytes.len()) {
                return Err((
                    "unavailable",
                    "Telemetry response exceeds the configured byte limit",
                ));
            }
            bytes.extend(chunk);
        }
        serde_json::from_slice(&bytes)
            .map_err(|_| ("unavailable", "Telemetry backend returned invalid JSON"))
    };
    let timer = Delay::from(Duration::from_millis(c.timeout_ms));
    futures_util::pin_mut!(fetch, timer);
    let result = match select(fetch, timer).await {
        Either::Left((result, _)) => result,
        Either::Right(_) => Err(unavailable()),
    };
    controller.abort();
    result
}
